cmd/monitor/main.go

```diff
@@ -0,0 +1,218 @@
+			}
+		} else {
+			glog.V(2).Infof("Polling: no new data found; tree size is still %d", st.LatestConsistent.Size)
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// Go around the loop again.
+		}
+	}
+}
+
+// Monitor verifiably checks inclusion of all leaves in a range, and then passes the
+// parsed FirmwareRelease to a handler.
+type Monitor struct {
+	st               client.LogStateTracker
+	stateFile        string
+	releaseVerifiers note.Verifiers
+	handler          func(uint64, api.FirmwareRelease) error
+}
+
+// From checks the leaves from `start` up to the checkpoint from the state tracker.
+// Upon reaching the end of the leaves, the checkpoint is persisted in the state file.
+func (m *Monitor) From(start uint64) error {
+	fromCP := m.st.LatestConsistent
+	pb, err := client.NewProofBuilder(fromCP, m.st.Hasher.HashChildren, m.st.Fetcher)
+
+	if err != nil {
+		return fmt.Errorf("failed to construct proof builder: %v", err)
+	}
+	for i := start; i < fromCP.Size; i++ {
+		rawLeaf, err := client.GetLeaf(m.st.Fetcher, i)
+		if err != nil {
+			return fmt.Errorf("failed to get leaf at index %d: %v", i, err)
+		}
+		hash := m.st.Hasher.HashLeaf(rawLeaf)
+		ip, err := pb.InclusionProof(i)
+		if err != nil {
+			return fmt.Errorf("failed to get inclusion proof for index %d: %v", i, err)
+		}
+		wantRoot, err := m.st.Verifier.RootFromInclusionProof(int64(i), int64(fromCP.Size), ip, hash)
+		if err != nil {
+			return fmt.Errorf("failed to calculate root from inclusion proof at index %d: %v", i, err)
+		}
+		if !bytes.Equal(wantRoot, fromCP.Hash) {
+			return fmt.Errorf("failed to verify inclusion proof for leaf %d.\ncheckpoint: %x\ncalculated: %x",
+				i, fromCP.Hash, wantRoot)
+		}
+
+		// XXX: this is unexpected. the hash is over the data with the trailing newline but we need to remove it
+		// in order to make this data into a valid note.
+		trimmedLeaf := strings.TrimSuffix(string(rawLeaf), "\n")
+		releaseNote, err := note.Open([]byte(trimmedLeaf), m.releaseVerifiers)
+		if err != nil {
+			return fmt.Errorf("failed to open leaf note at index %d: %v", i, err)
+		}
+
+		var release api.FirmwareRelease
+		if err := json.Unmarshal([]byte(releaseNote.Text), &release); err != nil {
+			return fmt.Errorf("failed to unmarshal release at index %d: %w", i, err)
+		}
+		if err := m.handler(i, release); err != nil {
+			return fmt.Errorf("handler(): %w", err)
+		}
+	}
+	return ioutil.WriteFile(m.stateFile, m.st.LatestConsistentRaw, 0644)
+}
+
+// stateTrackerFromFlags constructs a state tracker based on the flags provided to the main invocation.
+// The checkpoint returned will be the checkpoint representing this monitor's view of the log history.
+// A boolean is returned that is true iff the checkpoint was fetched from the log to initialize state.
+func stateTrackerFromFlags() (client.LogStateTracker, bool, error) {
+	if len(*stateFile) == 0 {
+		glog.Exit("--state_file required")
```

Comment:
nit: could return an error here and keep all the `Exitf` calls in the main func?

Reply:
Yeah that's better.